File: sentimental/caesar/caesar.py
# run program by typing 'python caesar.py' and an int that will serve as a key
# program will prompt for a message to be encrypted (plaintext)
# and outputs an encrypted version of that message (cipertext)
# the plaintext is shifted based off the numerical key given in the command line

import cs50
import sys

def main():
    # check that user provided a key in cmd line argument
    if len(sys.argv) != 2:
        print("Provide a key after running the program")
        exit(1)
# 1 get the key, cast key to integer
    key = int(sys.argv[1])
# 2 get the plaintext
    plaintext = cs50.get_string("Message to Encrypt: ")
    # initialize variable to hold ciphertext
    ciphertext = []
    print("ciphertext: ", end="")

# check characters for alpha + case, preserving case
# 3 once checked, encypher the text
    for x in range(len(plaintext)): # loop through plaintext string
        if plaintext[x].isalpha(): # check char is alpha
            if plaintext[x].isupper(): # check char is uppercase
                charIndex = ord(plaintext[x]) - 65 # set uppercase A-Z index to 0, by subtract 65
                ciphertext = chr((charIndex + key) % 26 + 65) # shift the char by they key, re-adding 65
                print(ciphertext, end="") # print shifted uppercase chars
            elif plaintext[x].islower(): # if lowercase
                charIndex = ord(plaintext[x]) - 97 # set lowercase a-z index to 0, by subtracting 97
                ciphertext = chr((charIndex + key) % 26 + 97) # shift the char by the key, re-adding 97
                print(ciphertext, end="") # print shifted lowercase chars
        else: # if not alpha, rpint as is
            ciphertext = plaintext[x]
            print(ciphertext, end="")
    # 4 print completed ciphertext on a new line
    print()

if __name__ == "__main__":
    main()

#Python Built In Functions Used

# ord(c)
# Given a string representing one Unicode character,
# return an integer representing the Unicode code point of that character.
# For example, ord('a') returns the integer 97 and ord('€') (Euro sign) returns 8364.
# This is the inverse of chr().


# chr(i)
# Return the string representing a character whose Unicode code point is the integer i.
# For example, chr(97) returns the string 'a', while chr(8364) returns the string '€'.
# This is the inverse of ord().

File: sentimental/caesar/caesar.c
// program will prompt for a message to be encrypted (plaintext)
// and outputs an encrypted version of that message (cipertext)
// the plaintext is shifted based off the numerical key given in the command line

#include<stdio.h> // printf
#include<cs50.h> // get_string
#include <string.h> // strlen
#include <ctype.h> // isupper, islower
#include <stdlib.h> // atoi

int main(int argc, string argv[])
{
    //if the command line argument isn't ./caesar #
    if (argc != 2)
    {
        printf("Error: must include a numerical key after executing program\n"); // tell the user to include a key
        return 1; // return the error code
    }

    // the key from command line comes through as a string, so atoi converts it to an int
    int keyAsInt = atoi(argv[1]); //the key is how many places to shift the chars

    string plaintext = get_string("Message to Encrypt: ");

    printf("ciphertext: "); // have to print "ciphertext" before the shifting process below

    for (int i = 0, n = strlen(plaintext); i < n; i++) //iterate through the plaintext string message
    {
        int letter; //holds the indexed uppercase + lowercase letters as an int
        char ciphertext; //holds the shifted letters

        if (isalpha(plaintext[i])) //if the chars are letters
        {
            if (isupper(plaintext[i])) //if the chars are letters and uppercase
            {
                letter = plaintext[i] - 'A'; //set uppercase A-Z index to 0 by subtracting 65 from all letters

                while (letter + keyAsInt > 25) // if the ASCII # for the char goes over 25 (the full uppercase alphabet)
                {
                    letter -= 26; //subtract when the limit of 26 is reached
                }
                letter += 'A'; //
                ciphertext = letter + keyAsInt; //take the indexed uppercase letter, shift by key# and set = to ciphertext
                printf("%c", ciphertext); //print each shifted char
            }
            else if (islower(plaintext[i])) //or if the chars are letters and lowercase
            {
                letter = plaintext[i] - 'a'; //set lowercase a-z index to 0 by subtracting 97 from all letters

                while (letter + keyAsInt > 25) // while the char's ASCII# is more than 25
                {
                    letter -= 26; //go back down 1 at a time or wrap back around
                }
                letter += 'a';
                ciphertext = letter + keyAsInt; //take the indexed lowercase letter, shift by key# and set = to ciphertext
                printf("%c", ciphertext); //print each shifted char
            }
        }
        else //if any of the the plaintext is not letters(uppercase or lowercase), print them as is - including spaces, symbols, numbers
        {
            ciphertext = plaintext[i]; //set the shifted message letters = ciphertext
            printf("%c", ciphertext); //print the cipher as chars
        }
    }
    printf("\n"); //print the final result on a new line
}